### ISSUE TITLE
Feature Request: Show DB Manager on Dock Icon Click (macOS)

### ISSUE THREAD

**Issue by dan-el** [strongbox-password-safe/Strongbox#584]
Currently, when Strongbox is running, and all app windows (database manager, any databases) are closed, a click on the Dock Icon does not bring up any part of the UI (database manager, "launch at startup" database, etc.). Instead we have to click the Dock Icon (so Strongbox is the active app) and then launch the database manager from the menu bar, from which in turn we can launch a database.

Can we have an advanced setting that allows us set what happens on clicking the Strongbox Dock Icon while all app windows are closed, but Strongbox is still running? I see two options: Open up the database manager or open up the "launch at startup" database.

Alternatively Strongbox could auto-quit when no app windows are open anymore.

Interestingly enough, when all app windows are closed, having the "Hide Dock Icon when all Miniturized or Closed" setting active leads a state of the App, where Strongbox is still running, but the Dock Icon is inactive (little dot disappeared), a click on the Dock Icon won't do anything, and Strongbox needs to be force quitted using Activity Monitor.

**Comment by strongbox-mark** [comment]
I think the best thing to do here is to launch the Databases Manager, however I'm having trouble reproducing the issue. If you restart Strongbox, are you seeing the issue straight away, or does it take some time to come into effect? 

I'm not seeing it on this side.

For the second issue, could you open another Github issue for that? Hard to manage 2 separate issue (but maybe related issues). Thanks!

**Comment by strongbox-mark** [comment]
I see the issue(s) thanks! 

So, step 5) is the main issue, nothing happens when you click on the Dock icon, so we need to look at that, I know you'd like to auto launch startup databases but I think that could be confusing/non intuitive, Databases Manager is my preference here. 

Quit when all windows closed could work but it would kill the menu bar/tray icon process too, so they would have to be mutually exclusive options I think.

Alternatively there is the option to hide the dock icon when all windows are closed/miniaturised if you're using the menu bar icon, which you might prefer. 

Thanks for the report!

**Comment by dan-el** [comment]
> So, step 5) is the main issue, nothing happens when you click on the Dock icon, so we need to look at that, I know you'd like to auto launch startup databases but I think that could be confusing/non intuitive, Databases Manager is my preference here.

If I have auto-launch database(s) set up I'd expect them to launch on a Dock icon click. Maybe you can make the "click action" follow the user's settings. Launch db manager if no auto-launch databases are set up, launch database(s) if auto-launch is set up. Otherwise it would feel rather unintuitive to set up auto-launch in the first place...

> Alternatively there is the option to hide the dock icon when all windows are closed/miniaturised if you're using the menu bar icon, which you might prefer.

That is the second issue I mentioned in my initial post. I will open up a separate issue now and then edit this comment to link the issues. TLDR: That option only hides the little dot next to the Dock icon but the Dock icon still doesn't do anything on-click.

**Comment by strongbox-mark** [comment]
TLDR: That option only hides the little dot next to the Dock icon but the Dock icon still doesn't do anything on-click.

I think this is because you have the Dock icon option "Keep in Dock" set for Strongbox, so it will always show whether Strongbox is running or not, Strongbox can't hide Dock icons that are set to "Keep in Dock". So, it seems maybe it's just the original bug.

**Comment by dan-el** [comment]
Here is the reference to #586

> I think this is because you have the Dock icon option "Keep in Dock" set for Strongbox, so it will always show whether Strongbox is running or not, Strongbox can't hide Dock icons that are set to "Keep in Dock". So, it seems maybe it's just the original bug.

Yes, I always have Strongbox pinned to the Dock since it is one of the most important apps on my Mac :-)
Maybe one of the two issues can be closed aas duplicate then.

**Comment by strongbox-mark** [comment]
OK, thanks @dan-el - Will investigate when we get around to this

**Comment by strongbox-mark** [comment]
Hi @dan-el - I believe this should be fixed for you now in 1.16.12.

Your Startup Databases will launch if you have them configured, otherwise the Databases Manager will be shown.

We've also added a "Quit Strongbox when all closed" preference so that the whole process terminates completely if you prefer that (obviously this doesn't work if you're using the Tray/Menu Bar Icon).

**Comment by dan-el** [comment]
Hi Mark,
I just tested this with Strongbox 1.16.12 and it seems it's not yet resolved. I still see the little dot next to the pinned Strongbox Icon in the Dock and nothing happens when I click it - regardless of whether the Menu Bar Icon is on or off.
I am on an M1 Pro MacBook on macOS 12.2.1

**Comment by strongbox-mark** [comment]
OK, it's strange, I can't reproduce this issue using just the below steps:

Steps to reproduce:

    Ensure the following Strongbox options are set:
    -> Advanced -> Show Databases Manager When All Windows are Closed - UNCHECKED
    -> Advanced -> Hide Databases Manager after Launching Database - CHECKED
    Open Strongbox and launch a database (regular Strongbox usage)
    Close Strongbox using the red X in the top left corner.
    -> Now we closed the last active app window. No database is launched, no databases manager window is open.
    The little dot in the Dock next to the Strongbox icon is still there, indicating Strongbox is still running.
    Click on the Strongbox icon in the Dock >> Nothing happens. That is the problem.

Is there anything else I should be looking for or setting?

**Comment by dan-el** [comment]
Yes, there is more. For some reason I only mentioned this in #586 
> Also requires -> General -> Show in System Menu Bar - CHECKED

and
> EDIT: I have Strongbox pinned to the Dock.

The more I think about these 2 issues (#584 and #586) the more I start to think that the problem lies within the fact that the "Show in System Menu Bar" option runs a) without an active app window but b) in same process as the regular app. That's why macOS sees the app as "active" (little dot next to the pinned Strongbox icon in the Dock) even when there is no regular app window.

Maybe another approach is to move the menu bar icon to a separate process that does not "belong to the dock icon" (in lack of the proper terms) - that would clear the "no active window but app still running" underlying issue maybe?

**Comment by georgesnow** [comment]
So this is purely for context. 

 correct if an app is running and has window(s). If the app process is “regular” it will show the dot under the dock icon. And when there is window available the dock icon will be displayed.

When most apps display as only “menu bar apps”. The process is transformed or modified  to what’s called an accessory app. Meaning it can have windows, but no dock icon. Some apps do you use a helper app and protocols like xpc services. that’s really a developer choice and what need the app is attempting fulfill of course.


this is just some insight if you were interested.

**Comment by dan-el** [comment]
@georgesnow the behaviour I see is that I have an active Dock icon (the indicator dot) even if only the Strongbox menu bar icon is active. And it doesn't matter if the icon is pinned to the Dock or not for this to happen. And if I choose to "Hide Dock Icon when all miniaturzied or closed" with the icon NOT pinned the dock icon also becomes unresponsive and Strongbox also won't launch from Spotlight Search - it can only be opened by choosing "Show Strongbox" from the menu bar icon.

Now we got #584 and #586 mixed up again :-)

**Comment by georgesnow** [comment]
I am nearly certain this is related to activation policy being set as prohibited.


I believe I can reliably recreate this issue. When the app is transformed to prohibited. it can’t have a dock icon therefore it won’t respond to any click actions. This is sort of an artifact of creating the “minimize to dock” effect the keepass proper (windows) has. And the timing of the dock icon appearing and hiding and when the minimizing a of window occurs.

this is a best guess from messing this personally when coding. But again I am just providing insight based on expereince
 



related to these issues:
#532 

https://github.com/keepassxreboot/keepassxc/pull/6575

**Comment by strongbox-mark** [comment]
Hi @dan-el - Could you check if this behaviour has been improved by 1.16.13?

**Comment by dan-el** [comment]
Hi @strongbox-mark - I reactivated the menu bar icon and the undesired behaviour remains unchanged with Strongbox pinned to the Dock.
Also when unpinning Strongbox from the Dock the app still doesn't open (e.g. from spotlight search) as long as the menu bar icon is still active - it appears as if the menu bar icon prevents any way of launching Strongbox other than through the menu bar icon menu itself.

**Comment by georgesnow** [comment]
So I ran quick test since I could recreate this. When the last window is closed (and dock is pinned). If you run the script below it returns that Strongbox is activation status is set to prohibited. Meaning the dock icon or launch the app from Applications folder, spotlight, et.. won't do anything as an app with this activation policy can't have windows or dock icon. That's why removing the app from dock doesn't effect it. The application is still technically running. 

However, status item (menubar) when you select a database "transforms" the application back to activation policy of regular thereby allowing windows to be created (and having a dock icon). 

The only difference in the behavior I can't recreate that @dan-el mentioned. The dock icon shows the indicator when it's running in this broken state. I can recreate all the behaviors except that one.

This is all to say and to  be helpful. I believe the app would need to monitor for when the last window is about to close (which I believe there is an api for? hopefully). The process would need to be set back to regular at that point when the window is closed.
  

````swift
#!/usr/bin/swift

import AppKit

let bundleMatchIdentifier = "com.markmcguill.strongbox.mac.pro"

NSWorkspace.shared.runningApplications
        .filter { $0 != NSRunningApplication.current }
        .filter { $0.activationPolicy == .prohibited }
        .filter { $0.bundleIdentifier == bundleMatchIdentifier  }
        .forEach { print($0) }
````

**Comment by strongbox-mark** [comment]
Hi guys, forgive the delay here... 

Unfortunately, I'm afraid I'm very confused as to what this bug is about. To that end could I ask a couple of things:

@georgesnow - I appreciate your input and I've no doubt that activation policy is part of the issue here but I'd just like to keep this conversation between myself and @dan-el from here on out, so that we have some hope of resolving/refining what the issue might be without digressions. 

@dan-el There's 2 points I want to make before continuing:

1. The use of "Hide Dock Icon" with "Pinning Strongbox to the Dock" is unsupported and will stay that way. To me, that configuration doesn't make any sense. As far as I can see, there are 2 options that make sense:
   - You want Strongbox in the Dock then => (Hide Dock Icon OFF & Pin to Dock (as you prefer))
   - You don't want Strongbox in the Dock then => (Hide Dock Icon ON & PIN Dock Icon OFF)
2. Given 1 we really need a super detailed and narrow description of a single particular behaviour that isn't working as you expect. If you think there's anything that might be related, please open another issue for it.

From (another!) re-read of the above thread, my guess as to how to phrase the real issue is the following:

_"When I have closed all the Strongbox windows I still see Strongbox in the Dock with a little light dot under it indicating it is running. When I click on it, nothing happens, but I think it should launch Quick Launch databases or the DB Manager"_

Is that a fair assessment of the core of the issue? If so, I'm afraid I can't reproduce that issue, but I would consider it a bug if it is the case.

**Comment by georgesnow** [comment]
👍 @strongbox-mark  no worries

**Comment by dan-el** [comment]
Hi @strongbox-mark,

I'll try to describe the issue I am seeing as precisely as possible - also I did some further testing.

**TLDR:** Combining the option "Show in the System Menu Bar" with the option "Hide Dock Icon when all miniaturized or closed" seems to create "the real bug" which is that Strongbox can only be "launched" via the System Menu Bar icon, and not via regular means (Spotlight, Dock Icon (if present), clicking the App in finder). I think once this is resolved it won't matter whether users pin the App to their Dock or not.

See my "test protocol" here: https://gist.github.com/dan-el/df9f86f80e3f11d249c61fdfc8e5628b

Hope this helps.

**Comment by dan-el** [comment]
Thanks a lot for a) taking the time and b) clearing my mess of bug reports :-)
